CRM/Civicase/Hook/Post/CreateSalesOrderContribution.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+use Civi\Api4\CaseSalesOrder;
+use Civi\Api4\CaseSalesOrderContribution;
+
+/**
+ * Handles sales order contribution post processing.
+ */
+class CRM_Civicase_Hook_Post_CreateSalesOrderContribution {
+
+  /**
+   * Creates Sales Order Contribtution.
+   *
+   * @param string $op
+   *   The operation being performed.
+   * @param string $objectName
+   *   Object name.
+   * @param mixed $objectId
+   *   Object ID.
+   * @param object $objectRef
+   *   Object reference.
+   */
+  public function run($op, $objectName, $objectId, &$objectRef) {
+    $toBeInvoiced = CRM_Utils_Request::retrieve('to_be_invoiced', 'String');
+    $percentValue = CRM_Utils_Request::retrieve('percent_value', 'Float');
+    $salesOrderId = CRM_Utils_Request::retrieve('sales_order', 'Integer');
+    $salesOrderStatusId = CRM_Utils_Request::retrieve('sales_order_status_id', 'Integer');
+
+    if (!$this->shouldRun($op, $objectName, $salesOrderId)) {
+      return;
+    }
+
+    $transaction = CRM_Core_Transaction::create();
+    try {
+      CaseSalesOrderContribution::create()
+        ->addValue('case_sales_order_id', $salesOrderId)
+        ->addValue('to_be_invoiced', $toBeInvoiced)
+        ->addValue('percent_value', $percentValue)
+        ->addValue('contribution_id', $objectId)
+        ->execute();
+
+      CaseSalesOrder::update()
+        ->addWhere('id', '=', $salesOrderId)
+        ->addValue('status_id', $salesOrderStatusId)
+        ->execute();
+    }
+    catch (\Throwable $th) {
+      $transaction->rollback();
```

Comment:
you will need to notify the user about an error as well instead of silently rollbacking